supervisor/core/supervisor.go

```diff
@@ -0,0 +1,214 @@
+/*
+Supervisor provides the ability to manage Instances.
+*/
+package core
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"sync"
+)
+
+// Supervisor stores the information about started Instances.
+type Supervisor struct {
+	// termMutex is used to prevent Instances from
+	// start / restart / stop during Supervisor termination.
+	termMutex sync.RWMutex
+	// instMutex is used to work with the map of Instances.
+	instMutex sync.RWMutex
+	// instancesById is a map of running Instances.
+	instancesById map[int]*Instance
+	// cfg is a pointer to a Supervisor config.
+	cfg *Cfg
+	// lastId is an id of the last running Instance.
+	lastId int
+}
+
+// NewSupervisor creates a Supervisor.
+func NewSupervisor(cfg *Cfg) *Supervisor {
+	sv := new(Supervisor)
+	sv.instancesById = make(map[int]*Instance)
+	sv.cfg = cfg
+	return sv
+}
+
+// addInstance adds the Instance to the Supervisor map.
+// Return ID of the Instance.
+func (sv *Supervisor) addInstance(inst *Instance) int {
+	sv.instMutex.Lock()
+	defer sv.instMutex.Unlock()
+	sv.lastId++
+	id := sv.lastId
+	sv.instancesById[id] = inst
+	return id
+}
+
+// deleteInstance removes the Instance from the Supervisor map.
+func (sv *Supervisor) deleteInstance(id int) {
+	sv.instMutex.Lock()
+	defer sv.instMutex.Unlock()
+	delete(sv.instancesById, id)
+}
+
+// getInstance return a pointer to the Instance by id.
+func (sv *Supervisor) getInstance(id int) *Instance {
+	sv.instMutex.RLock()
+	defer sv.instMutex.RUnlock()
+	inst := sv.instancesById[id]
+	return inst
+}
+
+// getInstanceByPid return an ID and a pointer ro the Instance by pid.
+func (sv *Supervisor) getInstanceByPid(pid int) (int, *Instance) {
+	// Seems like this shouldn't be a popular method
+	// and a full scan can be used.
+	sv.instMutex.RLock()
+	defer sv.instMutex.RUnlock()
+	for id, inst := range sv.instancesById {
+		if inst.Cmd.Process.Pid == pid {
+			return id, inst
+		}
+	}
+	return 0, nil
+}
+
+// StartInstance starts a new Instance with the specified parameters.
+// On fail returns 0, error.
+func (sv *Supervisor) StartInstance(name string, env []string, restartable bool) (int, error) {
+	// When Supervisor is terminating, we will lock "termMutex"
+	// to prevent new instances from starting during Supervisor termination.
+	sv.termMutex.RLock()
+	defer sv.termMutex.RUnlock()
+
+	// Form the path to the instance and check is it exists.
+	if name == "" {
+		return 0, errors.New(`The instance name is empty.`)
+	}
+	instPath := path.Join(sv.cfg.InstancesDir, name+".lua")
```

Comment:
`fmt.Sprintf("%s.lua", name)`.

Maybe later we need to add `getInstancePath` function.

BTW, `instancePath` isn't clear, it can be instance work dir, for example. Maybe it's `instanceEntrypiont`?

Reply:
> `fmt.Sprintf("%s.lua", name)`.

I don't understand what advantages `fmt.Sprintf ("% s.lua ", name)` has over `name +". lua "`

> 
> Maybe later we need to add `getInstancePath` function.

I don't think so. This is actually a configuration parameter.

> 
> BTW, `instancePath` isn't clear, it can be instance work dir, for example. Maybe it's `instanceEntrypiont`?

This is clear now (`tarantoolctl` legacy). I remember thinking about a symlink to a directory, but while it is not relevant.